frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1408 @@
+		/// migration, and they only wish for their transaction to do anything, if this assumption
+		/// holds. In case `witness_task` does not match, the transaction fails.
+		///
+		/// Based on the documentation of [`MigrationTask::migrate_until_exhaustion`], the
+		/// recommended way of doing this is to pass a `limit` that only bounds `count`, as the
+		/// `size` limit can always be overwritten.
+		#[pallet::weight(
+			// the migration process
+			Pallet::<T>::dynamic_weight(limits.item, * real_size_upper)
+			// rest of the operations, like deposit etc.
+			+ T::WeightInfo::continue_migrate()
+		)]
+		pub fn continue_migrate(
+			origin: OriginFor<T>,
+			limits: MigrationLimits,
+			real_size_upper: u32,
+			witness_task: MigrationTask<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let max_limits = T::SignedMigrationMaxLimits::get();
+			ensure!(
+				limits.size <= max_limits.size && limits.item <= max_limits.item,
+				"max signed limits not respected"
+			);
+
+			// ensure they can pay more than the fee.
+			let deposit = T::SignedDepositPerItem::get().saturating_mul(limits.item.into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut task = Self::migration_process();
+			ensure!(
+				task == witness_task,
+				DispatchErrorWithPostInfo {
+					error: "wrong witness".into(),
+					post_info: PostDispatchInfo {
+						actual_weight: Some(T::WeightInfo::continue_migrate_wrong_witness()),
+						pays_fee: Pays::Yes
+					}
+				}
+			);
+			task.migrate_until_exhaustion(limits);
+
+			// ensure that the migration witness data was correct.
+			if real_size_upper < task.dyn_size {
+				// let the imbalance burn.
+				let (_imbalance, _remainder) = T::Currency::slash(&who, deposit);
+				debug_assert!(_remainder.is_zero());
+				return Err("wrong witness data".into())
+			}
+
+			Self::deposit_event(Event::<T>::Migrated {
+				top: task.dyn_top_items,
+				child: task.dyn_child_items,
+				compute: MigrationCompute::Signed,
+			});
+
+			let actual_weight = Some(
+				Pallet::<T>::dynamic_weight(limits.item, task.dyn_size) +
+					T::WeightInfo::continue_migrate(),
+			);
+			MigrationProcess::<T>::put(task);
+			let pays = Pays::No;
+
+			Ok((actual_weight, pays).into())
+		}
+
+		/// Migrate the list of top keys by iterating each of them one by one.
+		///
+		/// This does not affect the global migration process tracker ([`MigrationProcess`]), and
+		/// should only be used in case any keys are leftover due to a bug.
+		#[pallet::weight(
+			T::WeightInfo::migrate_custom_top_success()
+				.max(T::WeightInfo::migrate_custom_top_fail())
+			.saturating_add(
+				Pallet::<T>::dynamic_weight(keys.len() as u32, *witness_size)
+			)
+		)]
+		pub fn migrate_custom_top(
+			origin: OriginFor<T>,
+			keys: Vec<Vec<u8>>,
+			witness_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
```

Comment:
won't that transaction be discarded by the transaction pool at some point? Isn't this almost the same as sending a `remark(10mb_data)` continuously to the any chain?

Reply:
@bkchr WDYT?

Reply:
The transaction pool will not be able to reject this, because it doesn't know what happens on execution.

But yeah, this is some possible attack vector.

Reply:
I don't have a good alternative, only think I am currently thinking of is going back to the inherent approach (like https://github.com/cheme/substrate/blob/4c42b43d0cd432df1244e4df899b340124cd801d/client/api/src/state_migration_inherents.rs and https://github.com/cheme/substrate/blob/4c42b43d0cd432df1244e4df899b340124cd801d/frame/state-migrate-0-to-1/src/lib.rs#L207), and keep only the 'migrate_custom' extrinsic for a single key/value at max weight cost.

The issue iirc is that we need to define an inherent just for it. The way I register the inherent was not at all modular https://github.com/cheme/substrate/blob/4c42b43d0cd432df1244e4df899b340124cd801d/bin/node/cli/src/service.rs#L425 and requiring https://github.com/cheme/substrate/blob/4c42b43d0cd432df1244e4df899b340124cd801d/frame/state-migrate-0-to-1/src/lib.rs#L243 , maybe it is possible to do it differently.

Also thinking with 'migrate_custom' limited to a single key at max weight, and 'continue_migrate' kept as it is, the ddos will be limited to certain points in time (since migration progresses); may be 'manageable', but I am not an optimistic person.

Reply:
@cheme I think the solution to both of these that we advice chains (including relay chains) to limit who can call this. In the code, we make it configurable, and allow each runtime to define its own `EnsureOrigin`. Agree?

Reply:
Yes we can do it this way. With possibly different `EnsureOrigin` for `migrate_custom` and `continue_migrate` (I can think some may take the risk of running `continue_migrate` without a restricted `EnsureOrigin`).